File: app/src/test/java/com/abishov/hexocat/home/trending/TrendingPresenterUnitTests.java
package com.abishov.hexocat.home.trending;

import com.abishov.hexocat.commons.schedulers.TrampolineSchedulersProvider;
import com.abishov.hexocat.github.Repository;
import com.abishov.hexocat.github.User;
import com.abishov.hexocat.github.filters.SearchQuery;
import com.abishov.hexocat.home.repository.RepositoryViewModel;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.Answers;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.threeten.bp.LocalDate;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import io.reactivex.subjects.BehaviorSubject;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(JUnit4.class)
public class TrendingPresenterUnitTests {

    @Mock(answer = Answers.RETURNS_DEEP_STUBS)
    private TrendingContract.View trendingView;

    @Mock
    private TrendingRepository trendingRepository;

    @Captor
    private ArgumentCaptor<TrendingViewState> repositoriesConsumer;

    // Faking behaviour of the TrendingRepository and TrendingContract.View.
    private BehaviorSubject<SearchQuery> viewQueries;
    private BehaviorSubject<List<Repository>> listResults;

    // Some dummy data.
    private List<Repository> repositories;

    // ViewModels corresponding to dummy data.
    private List<RepositoryViewModel> repositoryViewModels;

    private TrendingPresenter trendingPresenter;
    private SearchQuery searchQuery;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);

        viewQueries = BehaviorSubject.create();
        listResults = BehaviorSubject.create();
        trendingPresenter = new TrendingPresenter(
                new TrampolineSchedulersProvider(), trendingRepository);

        User owner = User.create("test_login", "test_html_url", "test_avatar_url");
        repositories = Arrays.asList(
                Repository.create("test_repository_one",
                        "test_html_url_one", 5, 10, "test_description_one", owner),
                Repository.create("test_repository_two",
                        "test_html_url_two", 7, 11, "test_description_two", owner));
        repositoryViewModels = Arrays.asList(
                RepositoryViewModel.create("test_repository_one", "test_description_one",
                        "5", "10", "test_avatar_url", "test_login"),
                RepositoryViewModel.create("test_repository_two", "test_description_two",
                        "7", "11", "test_avatar_url", "test_login"));

        searchQuery = new SearchQuery.Builder()
                .createdSince(LocalDate.parse("2017-08-10"))
                .build();

        when(trendingView.searchQueries()).thenReturn(viewQueries);
        when(trendingRepository.trendingRepositories(searchQuery)).thenReturn(listResults);
    }

    @Test
    public void presenterMustPropagateCorrectStatesOnSuccess() throws Exception {
        trendingPresenter.onAttach(trendingView);
        assertThat(viewQueries.hasObservers()).isTrue();

        viewQueries.onNext(searchQuery);
        listResults.onNext(repositories);
        listResults.onComplete();

        verify(trendingView.bindTo(), times(2))
                .accept(repositoriesConsumer.capture());

        TrendingViewState viewStateProgress =
                repositoriesConsumer.getAllValues().get(0);

        assertThat(viewStateProgress.isSuccess()).isFalse();
        assertThat(viewStateProgress.isInProgress()).isTrue();
        assertThat(viewStateProgress.isFailure()).isFalse();
        assertThat(viewStateProgress.isIdle()).isFalse();

        TrendingViewState viewStateSuccess =
                repositoriesConsumer.getAllValues().get(1);

        assertThat(viewStateSuccess.isSuccess()).isTrue();
        assertThat(viewStateSuccess.isInProgress()).isFalse();
        assertThat(viewStateSuccess.isFailure()).isFalse();
        assertThat(viewStateSuccess.isIdle()).isFalse();
        assertThat(viewStateSuccess.items()).isEqualTo(repositoryViewModels);
    }

    @Test
    public void presenterMustPropagateCorrectStatesOnFailure() throws Exception {
        trendingPresenter.onAttach(trendingView);
        assertThat(viewQueries.hasObservers()).isTrue();

        viewQueries.onNext(searchQuery);
        listResults.onError(new Throwable("test_message"));
        listResults.onComplete();

        verify(trendingView.bindTo(), times(2))
                .accept(repositoriesConsumer.capture());

        TrendingViewState viewStateProgress =
                repositoriesConsumer.getAllValues().get(0);

        assertThat(viewStateProgress.isSuccess()).isFalse();
        assertThat(viewStateProgress.isInProgress()).isTrue();
        assertThat(viewStateProgress.isFailure()).isFalse();
        assertThat(viewStateProgress.isIdle()).isFalse();

        TrendingViewState viewStateSuccess =
                repositoriesConsumer.getAllValues().get(1);

        assertThat(viewStateSuccess.isSuccess()).isFalse();
        assertThat(viewStateSuccess.isInProgress()).isFalse();
        assertThat(viewStateSuccess.isFailure()).isTrue();
        assertThat(viewStateSuccess.isIdle()).isFalse();
        assertThat(viewStateSuccess.items()).isEmpty();
    }

    @Test
    public void presenterMustPropagateCorrectStatesWhenNoItems() throws Exception {
        trendingPresenter.onAttach(trendingView);
        assertThat(viewQueries.hasObservers()).isTrue();

        viewQueries.onNext(searchQuery);
        listResults.onNext(new ArrayList<>());
        listResults.onComplete();

        verify(trendingView.bindTo(), times(2))
                .accept(repositoriesConsumer.capture());

        TrendingViewState viewStateProgress = repositoriesConsumer.getAllValues().get(0);

        assertThat(viewStateProgress.isSuccess()).isFalse();
        assertThat(viewStateProgress.isInProgress()).isTrue();
        assertThat(viewStateProgress.isFailure()).isFalse();
        assertThat(viewStateProgress.isIdle()).isFalse();

        TrendingViewState viewStateSuccess = repositoriesConsumer.getAllValues().get(1);

        assertThat(viewStateSuccess.isSuccess()).isTrue();
        assertThat(viewStateSuccess.isInProgress()).isFalse();
        assertThat(viewStateSuccess.isFailure()).isFalse();
        assertThat(viewStateSuccess.isIdle()).isFalse();
        assertThat(viewStateSuccess.items()).isEmpty();
    }

    @Test
    public void presenterMustUnsubscribeFromViewOnDetach() {
        assertThat(listResults.hasObservers()).isFalse();
        assertThat(viewQueries.hasObservers()).isFalse();

        trendingPresenter.onAttach(trendingView);
        assertThat(listResults.hasObservers()).isFalse();
        assertThat(viewQueries.hasObservers()).isTrue();

        trendingPresenter.onDetach();
        assertThat(listResults.hasObservers()).isFalse();
        assertThat(viewQueries.hasObservers()).isFalse();
    }
}

File: app/src/test/java/com/abishov/hexocat/home/trending/TrendingPresenterUnitTests.kt
package com.abishov.hexocat.home.trending

import android.net.Uri
import androidx.test.ext.junit.runners.AndroidJUnit4
import assertk.assertThat
import assertk.assertions.*
import com.abishov.hexocat.common.schedulers.TrampolineSchedulersProvider
import com.abishov.hexocat.components.*
import com.abishov.hexocat.github.filters.SearchQuery
import com.github.TrendingRepositoriesQuery.*
import com.nhaarman.mockitokotlin2.whenever
import io.reactivex.subjects.BehaviorSubject
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.*
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.threeten.bp.LocalDate
import java.util.*

@RunWith(AndroidJUnit4::class)
class TrendingPresenterUnitTests {

  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
  private lateinit var trendingView: TrendingContract.View

  @Mock
  private lateinit var trendingService: TrendingService

  @Captor
  private lateinit var repositoriesConsumer: ArgumentCaptor<TrendingViewState>

  private lateinit var viewQueries: BehaviorSubject<SearchQuery>
  private lateinit var listResults: BehaviorSubject<List<AsRepository>>
  private lateinit var searchQuery: SearchQuery

  private lateinit var repositories: List<AsRepository>
  private lateinit var repositoryViewModels: List<RepositoryViewModel>

  private lateinit var trendingPresenter: TrendingPresenter

  @Before
  fun setUp() {
    MockitoAnnotations.initMocks(this)

    viewQueries = BehaviorSubject.create()
    listResults = BehaviorSubject.create()

    trendingPresenter = TrendingPresenter(
      TrampolineSchedulersProvider(), trendingService
    )

    val owner = Owner(
      id = "test_owner_id",
      login = "test_login",
      avatarUrl = Uri.parse("http://github.com/test_avatar_url")
    )

    val topics = RepositoryTopics(
      topics = listOf(
        Topic(topic = Topic1(name = "test_topic_1")),
        Topic(topic = Topic1(name = "test_topic_2"))
      )
    )

    repositories = listOf(
      AsRepository(
        name = "test_repository_one",
        url = Uri.parse("http://github.com/test_html_url_one"),
        stargazerCount = 10,
        description = "test_description_one",
        openGraphImageUrl = Uri.parse("http://github.com/test_banner_url_one"),
        usesCustomOpenGraphImage = true,
        owner = owner,
        repositoryTopics = topics,
        primaryLanguage = PrimaryLanguage(name = "test_lang_1", color = "test_lang_color_1"),
        mentionableUsers = MentionableUsers(contributors = listOf(), totalCount = 0)
      ),
      AsRepository(
        name = "test_repository_two",
        url = Uri.parse("http://github.com/test_html_url_two"),
        stargazerCount = 11,
        description = "test_description_two",
        openGraphImageUrl = Uri.parse("http://github.com/test_banner_url_two"),
        usesCustomOpenGraphImage = true,
        owner = owner,
        repositoryTopics = RepositoryTopics(topics = listOf()),
        primaryLanguage = PrimaryLanguage(name = "test_lang_2", color = "test_lang_color_2"),
        mentionableUsers = MentionableUsers(contributors = listOf(), totalCount = 0)
      )
    )

    repositoryViewModels = listOf(
      RepositoryViewModel(
        name = "test_repository_one",
        description = "test_description_one",
        stars = 10,
        owner = OwnerViewModel(
          avatarUrl = Uri.parse("http://github.com/test_avatar_url"),
          login = "test_login",
          id = "test_owner_id"
        ),
        url = Uri.parse("http://github.com/test_html_url_one"),
        bannerUrl = Uri.parse("http://github.com/test_banner_url_one"),
        topics = listOf(
          TopicViewModel(name = "test_topic_1"),
          TopicViewModel(name = "test_topic_2")
        ),
        primaryLanguage = LanguageViewModel(name = "test_lang_1", color = "test_lang_color_1"),
        mentionableUsers = MentionableUsersViewModel(listOf(), 0)
      ),
      RepositoryViewModel(
        name = "test_repository_two",
        description = "test_description_two",
        stars = 11,
        owner = OwnerViewModel(
          avatarUrl = Uri.parse("http://github.com/test_avatar_url"),
          login = "test_login",
          id = "test_owner_id",
        ),
        url = Uri.parse("http://github.com/test_html_url_two"),
        bannerUrl = Uri.parse("http://github.com/test_banner_url_two"),
        topics = listOf(),
        primaryLanguage = LanguageViewModel(name = "test_lang_2", color = "test_lang_color_2"),
        mentionableUsers = MentionableUsersViewModel(listOf(), 0)
      )
    )

    searchQuery = SearchQuery(LocalDate.parse("2017-08-10"))
    whenever(trendingView.searchQueries()).thenReturn(viewQueries)
    whenever(trendingService.search(searchQuery, 32)).thenReturn(
      listResults
    )
  }

  @Test
  @Throws(Exception::class)
  fun `presenter must propagate correct states on success`() {
    trendingPresenter.onAttach(trendingView)
    assertThat(viewQueries.hasObservers()).isTrue()

    viewQueries.onNext(searchQuery)
    listResults.onNext(repositories)
    listResults.onComplete()

    verify(trendingView.bindTo(), times(2))
      .accept(repositoriesConsumer.capture())

    val viewStateProgress = repositoriesConsumer.allValues[0]
    assertThat(viewStateProgress).isInstanceOf(TrendingViewState.InProgress::class)

    val viewStateSuccess = repositoriesConsumer.allValues[1]
    assertThat(viewStateSuccess).isInstanceOf(TrendingViewState.Success::class)
    assertThat((viewStateSuccess as TrendingViewState.Success).items).isEqualTo(repositoryViewModels)
  }

  @Test
  @Throws(Exception::class)
  fun `presenter must propagate correct states on failure`() {
    trendingPresenter.onAttach(trendingView)
    assertThat(viewQueries.hasObservers()).isTrue()

    viewQueries.onNext(searchQuery)
    listResults.onError(Throwable("test_message"))
    listResults.onComplete()

    verify(trendingView.bindTo(), times(2))
      .accept(repositoriesConsumer.capture())

    val viewStateProgress = repositoriesConsumer.allValues[0]
    assertThat(viewStateProgress).isInstanceOf(TrendingViewState.InProgress::class)

    val viewStateFailure = repositoriesConsumer.allValues[1]
    assertThat(viewStateFailure).isInstanceOf(TrendingViewState.Failure::class)
  }

  @Test
  @Throws(Exception::class)
  fun `presenter must propagate correct states when no items`() {
    trendingPresenter.onAttach(trendingView)
    assertThat(viewQueries.hasObservers()).isTrue()

    viewQueries.onNext(searchQuery)
    listResults.onNext(ArrayList())
    listResults.onComplete()

    verify(trendingView.bindTo(), times(2))
      .accept(repositoriesConsumer.capture())

    val viewStateProgress = repositoriesConsumer.allValues[0]
    assertThat(viewStateProgress).isInstanceOf(TrendingViewState.InProgress::class)

    val viewStateSuccess = repositoriesConsumer.allValues[1]
    assertThat(viewStateSuccess).isInstanceOf(TrendingViewState.Success::class)
    assertThat((viewStateSuccess as TrendingViewState.Success).items).isEmpty()
  }

  @Test
  fun `presenter must unsubscribe from view on detach`() {
    assertThat(listResults.hasObservers()).isFalse()
    assertThat(viewQueries.hasObservers()).isFalse()

    trendingPresenter.onAttach(trendingView)
    assertThat(listResults.hasObservers()).isFalse()
    assertThat(viewQueries.hasObservers()).isTrue()

    trendingPresenter.onDetach()
    assertThat(listResults.hasObservers()).isFalse()
    assertThat(viewQueries.hasObservers()).isFalse()
  }
}
